clang/lib/Frontend/InitPreprocessor.cpp

```diff
@@ -1558,6 +1558,13 @@ void clang::InitializePreprocessor(Preprocessor &PP,
                                    const PCHContainerReader &PCHContainerRdr,
                                    const FrontendOptions &FEOpts,
                                    const CodeGenOptions &CodeGenOpts) {
+
+  if (all_of(FEOpts.Inputs,
+             [](const FrontendInputFile &FI) { return FI.isPreprocessed(); })) {
+    PP.SetDisableMacroExpansion();
+    return;
+  }
```

Comment:
Did you consider doing that on a per-file basis in `FrontendAction::BeginSourceFile`?

Reply:
You're right, it makes more sense there. I've updated it to match.